apps/liteprotocoltester/statistics.nim

```diff
@@ -0,0 +1,173 @@
+  self.helper.prevArrivedAt = currentArrivedAt
+  self.helper.prevIndex = msg.index
+
+  inc(self.receivedMessages)
+
+  if self.minLatency.isZero or (latency < self.minLatency and latency > nanos(0)):
+    self.minLatency = latency
+  if latency > self.maxLatency:
+    self.maxLatency = latency
+
+  self.cummulativeLatency += latency
+
+proc addMessage*(
+    self: var PerPeerStatistics, peerId: string, msg: ProtocolTesterMessage
+) =
+  if not self.contains(peerId):
+    self[peerId] = Statistics.init()
+
+  discard catch:
+    self[peerId].addMessage(msg)
+
+proc lossCount*(self: Statistics): uint32 =
+  self.allMessageCount - self.receivedMessages
+
+proc averageLatency*(self: Statistics): Duration =
+  if self.receivedMessages == 0:
+    return nanos(0)
+  return self.cummulativeLatency div self.receivedMessages
+
+proc echoStat*(self: Statistics) =
+  let printable = catch:
+    """*----------------------------------------------------------------*
+|  Expected  |  Reveived  |    Loss    |  Misorder  |  Duplicate |
+|{self.allMessageCount:>11} |{self.receivedMessages:>11} |{self.lossCount():>11} |{self.misorderCount:>11} |{self.duplicateCount:>11} |
+*----------------------------------------------------------------*
+| Latency stat:                                                  |
+|    avg latency: {self.averageLatency():<47}|
+|    min latency: {self.maxLatency:<47}|
+|    max latency: {self.minLatency:<47}|
+*----------------------------------------------------------------*""".fmt()
+
+  if printable.isErr:
+    echo "Error while printing statistics"
+  else:
+    echo printable.get()
+
+proc jsonStat*(self: Statistics): string =
+  let json = catch:
+    """{{\"expected\":{self.allMessageCount},
+         \"received\": {self.receivedMessages},
+         \"loss\": {self.lossCount()},
+         \"misorder\": {self.misorderCount},
+         \"duplicate\": {self.duplicateCount},
+         \"latency\":
+           {{\"avg\": {self.averageLatency()},
+             \"min\": {self.minLatency},
+             \"max\": {self.maxLatency}
+           }}
+     }}""".fmt()
+  if json.isErr:
+    return "{\"result:\": \"" & json.error.msg & "\"}"
+
+  return json.get()
+
+proc echoStats*(self: var PerPeerStatistics) =
+  for peerId, stats in self.pairs:
+    let peerLine = catch:
+      "Receiver statistics from peer {peerId}".fmt()
+    if peerLine.isErr:
+      echo "Error while printing statistics"
+    else:
+      echo peerLine.get()
+      stats.echoStat()
+
+proc jsonStats*(self: var PerPeerStatistics): string =
+  try:
+    #!fmt: off
+    var json = "{\"statitics\": ["
+    for peerId, stats in self.pairs:
+      json.add("{{\"sender\": \"{peerId}\", \"stat\":".fmt())
+      json.add(stats.jsonStat())
+      json.add("}, ")
+    json.add("]}")
+    return json
+    #!fmt: on
+  except CatchableError:
+    return
+      "{\"result:\": \"Error while generating json stats: " & getCurrentExceptionMsg() &
+      "\"}"
```

Comment:
The `std/json` module is pretty handy for such cases.
f.e. https://github.com/waku-org/nwaku/blob/e7b0777d2599e9d33f407dca8de2b1791bdf8320/library/events/json_message_event.nim#L97


```suggestion
  return $(%*self)
```

Reply:
I checked it, but cannot do such easy shortcut unfortunately, that is more for key-value pairs or containers...